added back to repo

MikeLippincott opened the pull request.

d33bs commented on setup.py:
```diff
@@ -0,0 +1,26 @@
+    name="cytocherrypick",
+    version="0.1",
+    packages=find_packages(),
+    install_requires=[
```
Consider updating this line to reference the `requirements.txt` file to source this information from a single place (where otherwise certain updates may be forgotten in one or the other location). See the following [example from Pycytominer](https://github.com/cytomining/pycytominer/blob/c4de0a9a4fecbf1ad11872bb14c18d24d1b1851e/setup.py#L28C22-L28C35).

MikeLippincott replied:
addressed

d33bs commented on 0.Data_Preprocessing/scripts/0.generate_toy_data.py:
Consider moving this file to be located next to the data which it generates under the `tests` dir. Doing this will help demonstrate where the test is generated from.

MikeLippincott replied:
I moved the test_data generating file to the tests directory

d33bs commented on tests/data/test_data.csv:
Could this data be generated as part of a [`pytest fixture`](https://docs.pytest.org/en/latest/explanation/fixtures.html), making the sample data more dynamic in nature? Doing this might help adjust the test data over time without having to change the CSV in the repo. It also might open the project up to greater testing potential with randomized data (random for each test run).

MikeLippincott replied:
Also adding to an issue as I might address this later

d33bs commented on tmp.txt:
Consider removing this file if it is not used.

d33bs commented on tests/main_tests/median_test.py:
```diff
@@ -0,0 +1,25 @@
+from src.utils.helper_functions import find_median
+
+
+class TestGetMedian(unittest.TestCase):
```
Minor comment: consider matching the name of the function which is tested here, i.e. `find` vs `get`.

d33bs commented on README.md:
```diff
@@ -0,0 +1,27 @@
+# CytoCherryPick
+
+## Welcome to CytoCherryPick
+CytoCherryPick is an opensource sowftware that provides a set of tools for the selection of representative images from a set of images.
```
```suggestion
CytoCherryPick is an open source software that provides a set of tools for the selection of representative images from a set of images.
```

d33bs commented on README.md:
```diff
@@ -0,0 +1,27 @@
+CytoCherryPick is an opensource sowftware that provides a set of tools for the selection of representative images from a set of images.
+
+## Functions of CytoCherryPick
+* **Image selection**: CytoCherryPick will select a representative image from a set of images.
+* **Image Distribution Analysis**: Choose your own representative image and CytoCherryPick will graph where in the image set distribution your image is located.
```
Consider maintaining consistent capitalization for the bolded text at the beginning of each line in this list.

d33bs commented on README.md:
````diff
@@ -0,0 +1,27 @@
+$ cd CytoCherryPick
+```
+
+### Install the CytoCherryPick env
````
Consider completing the abbreviated `env` here as `environment` or defining it as an abbreviation somewhere before this line.

d33bs commented on 0.Data_Preprocessing/scripts/0.generate_toy_data.py:
```diff
@@ -0,0 +1,70 @@
+image_name = "image_1"
+for col in df.columns:
+    if col != "image_name":
+        sns.distplot(df[col], bins=bins, kde=True, hist=False, kde_kws={"shade": True})
+        plt.axvline(
+            df[df["image_name"] == image_name][col].values[0], color="r", linestyle="--"
+        )
```
Consider adding some comments here to describe what's happening (and what each line adds to the chart).

MikeLippincott replied:
Good point, added

d33bs commented on 0.Data_Preprocessing/scripts/0.generate_toy_data.py:
```diff
@@ -0,0 +1,70 @@
+# ---
+# jupyter:
+#   jupytext:
```
If it hasn't already been done, consider adding `jupytext` as a development dependency for this project.

d33bs commented on 0.Data_Preprocessing/scripts/0.generate_toy_data.py:
```diff
@@ -0,0 +1,70 @@
+image_name = "image_1"
+for col in df.columns:
+    if col != "image_name":
+        sns.distplot(df[col], bins=bins, kde=True, hist=False, kde_kws={"shade": True})
```
I noticed this warning in the executed notebook output about how `shade` is interpreted. Consider addressing the warning to help prepare this work for future `seaborn` implementations and to avoid code decay.

```text
`shade` is now deprecated in favor of `fill`; setting `fill=True`.
This will become an error in seaborn v0.14.0; please update your code.
```

d33bs commented on 0.Data_Preprocessing/notebooks/0.generate_toy_data.ipynb:
```diff
@@ -0,0 +1,442 @@
+                },
+                {
+                    "data": {
+                        "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkAAAAHFCAYAAAAaD0bAAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAABghUlEQVR4nO3deVxU9f4/8NfMwMywzbAPIIsIyiKKCopg7orapq3c+qZ108rbcjN/fe/Na93Ku3j7tlldtbyVXCsRW0y7aYqVC4IbgruIGyCyKwzrADPn94fFDUEDZPjMMK/n43Eej/HMmcPrTDHz5nM+i0ySJAlERERENkQuOgARERFRb2MBRERERDaHBRARE ...
```
Depending on how you plan to use notebooks and `jupytext` with this project, consider adding the following to your `pyproject.toml` file in order to tell `jupytext` how to pair notebook files throughout the project. This might need to be modified in order to retain the "scripts" sub-directory structure.

```toml
[tool.jupytext]
formats = "ipynb,py:percent"
```

MikeLippincott replied:
notebook is an artifact. I have been using them for trouble shooting but I dont't think I will have any soon.

MikeLippincott commented on 0.Data_Preprocessing/scripts/0.generate_toy_data.py:
```diff
@@ -0,0 +1,70 @@
+# set seed
+np.random.seed(40)
+# generate fake image data with 6 columns and 100 rows with image names as one of the columns and data from random gaussian data, geometric, poisson, binomial, exponential, uniform distributions
+df = pd.DataFrame(np.random.randn(100, 6), columns=list("ABCDEF"))
+# df['image_name'] = ['image_' + str(i) for i in range(100)]
+# df['geometric'] = np.random.geometric(p=0.5, size=100)
+# df['poisson'] = np.random.poisson(lam=1.0, size=100)
+# df['binomial'] = np.random.binomial(n=10, p=0.5, size=100)
+# df['exponential'] = np.random.exponential(scale=1.0, size=100)
```
great suggestion. I added this.

MikeLippincott commented on README.md:
````diff
@@ -0,0 +1,27 @@
+
+### Install the CytoCherryPick env
+``` bash
+$ mamba env install --file=env/env.yml
````
hmmm, I am not sure. This realm is new to me and we might need to talk more about how I can do this. I will be picking your brain later!

MikeLippincott commented on 0.Data_Preprocessing/scripts/0.generate_toy_data.py:
```diff
@@ -0,0 +1,70 @@
+# add image name column to the dataframe
+df["image_name"] = ["image_" + str(i) for i in range(100)]
+
+df.to_csv(test_data_output_path, index=False)
```
Probably, but this is just reproducible toy data to test things on, so a new index when loading this csv isn't bad?

d33bs replied:
@MikeLippincott thanks, I think this isn't a big deal for now, but it could be something to think about when it comes to data testing or future functionality. In cases where someone might have accidentally included the index (as a default when using `pd.to_csv()`) could it negatively effect the outcomes?

MikeLippincott commented on requirments.txt:
```diff
@@ -0,0 +1,3 @@
+pandas
+seaborn
+matplotlib
```
Added!

MikeLippincott commented on 0.Data_Preprocessing/scripts/0.generate_toy_data.py:
addressed and changed. Good point about the test vs toy data

MikeLippincott commented on CytoCherryPick.sh:
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+# call the main.py file with the following arguments to be parsed
+python Utils/main.py --file_path "data/test_data_output.csv" \
```
I am thinking of making this a python tool to go along with the cytomining suite. I am not sure if I need a CLI interface? I will make an issue and we can talk more later!

d33bs replied:
Sounds good! @MikeLippincott , food for thought as you think through design:
- ["Deep classes" (or functions)](https://youtu.be/bmSAYlu0NcY?si=J47hnUePJEObSmyi&t=756) highlights how you can balance value-add vs complexity.
- [Hyrums law](https://www.hyrumslaw.com/) highlights that we should be choosy about what our API includes (or excludes) as eventually someone might depend on it, sometimes in unexpected ways.

d33bs commented on tests/main_tests/test_data_check.py:
```diff
@@ -0,0 +1,45 @@
+        """
+
+        file_out_path = "tests/data/test_data.csv"
+        self.df = generate_test_data(file_out_path, 100)
```
Consider validating the row length of the test data as an additional test here.

MikeLippincott replied:
added function

d33bs commented on tests/main_tests/test_data_check.py:
```diff
@@ -0,0 +1,45 @@
+        self.test_data_check()
+        num_of_rows = 100
+        np.random.seed(0)
+        df = pd.DataFrame(
+            {
+                "feature1": np.random.rand(num_of_rows),
+                "feature2": np.random.rand(num_of_rows),
+                "feature3": np.random.normal(-3, 1, num_of_rows),
+                "image_name": ["image_" + str(i) for i in range(num_of_rows)],
+            }
+        )
```
Consider providing a code comment about why this dataframe has the structure it does.

MikeLippincott replied:
added

d33bs reviewed (changes requested):
Great job, thank you for addressing the comments! I left a few comments with this review for your consideration. I'm marking this as requesting changes due to the inclusion of the `.pyc` files. Please don't hesitate to let me know if you have any questions.

As a general comment with this PR, I'd recommend leaning towards smaller portions of additions to make for more distinct chunks of reviewable work. For example, environment management, testing, and initial package/modules all could be individual PR's. While the number of lines is relatively low, each of these ideas are reasonably complex and have many options.

MikeLippincott commented on tests/main_tests/__init__.py:
Moved file!

MikeLippincott commented on pyproject.toml:
```diff
@@ -1,8 +1,45 @@
+Homepage = "https://github.com/MikeLippincott/CytoCherryPick"
+
+[tool.hatch.version]
+path = "cytocherrypick/__init__.py"
```
File created

MikeLippincott commented on requirements.txt:
Yes, agreed! Thank you

MikeLippincott commented on pyproject.toml:
Added an issue an will open another PR for this

MikeLippincott replied:
On second thought, I changed to poetry in this PR as I couldn't build the package with hatch....

MikeLippincott commented on pyproject.toml:
```diff
@@ -1,8 +1,45 @@
+dynamic = ["version"]
+description = "A tool for visualizing a miscropscopy image's feature space."
+readme = "README.md"
+license = ""
```
Licensing filled out

MikeLippincott commented on src/utils/helper_functions/find_median.py:
```diff
@@ -0,0 +1,27 @@
+
+    Parameters
+    ----------
+    df : Pandas.DataFrame
```
good point, thank you for pointing this out

MikeLippincott commented on src/utils/helper_functions/generate_test_data.py:
```diff
@@ -0,0 +1,34 @@
+
+def generate_test_data(output_file_path: str, num_of_rows: int = 100):
+    # Define paths
+    test_output_path = output_file_path.split("/")[-1]
```
adjusted to fix the possible error

MikeLippincott commented on src/utils/helper_functions/generate_test_data.py:
```diff
@@ -0,0 +1,34 @@
+import seaborn as sns
+
+
+def generate_test_data(output_file_path: str, num_of_rows: int = 100):
```
Docstring added

MikeLippincott commented on src/cherrypick.py:
```diff
@@ -0,0 +1,50 @@
+import matplotlib.pyplot as plt
```
Docstring added

MikeLippincott commented on src/utils/helper_functions/generate_test_data.py:
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env python
```
I didn't know this, thank you for spreading the knowledge! 
Interesting that you call it a `hash-bang` I usually call it a `shebang`

d33bs replied:
@MikeLippincott - just a minor note here on the naming convention: I tend to prefer using the literal "hash == #" and "bang == !" version to avoid gender-specific connotations which could be inferred. Mostly the intent on my part here is towards doing no harm. I believe the origins of "shebang" might be obscure, so it's hard to say what the "she" part specifically came from or what it might imply.

MikeLippincott commented on src/utils/helper_functions/plot_distribution.py:
```diff
@@ -0,0 +1,51 @@
+import matplotlib.pyplot as plt
```
docstring added

MikeLippincott commented on tests/main_tests/test_data_check.py:
```diff
@@ -0,0 +1,45 @@
+        """
+        check to ensure that the test data is generated correctly
+
+        returns:
+            None
```
I also did not know this. Thank you for spreading the knowledge

MikeLippincott commented on tests/main_tests/test_data_check.py:
```diff
@@ -0,0 +1,45 @@
+        """
+
+        file_out_path = "tests/data/test_data.csv"
+        self.df = generate_test_data(file_out_path, 100)
```
Changed

MikeLippincott commented on tests/main_tests/test_data_check.py:
```diff
@@ -0,0 +1,45 @@
+        returns:
+            None
+        """
+        self.test_data_check()
```
Good point I created a new class

MikeLippincott commented on src/utils/helper_functions/generate_test_data.py:
```diff
@@ -0,0 +1,34 @@
+    test_output_path = pathlib.Path(output_file_path)
+
+    # set seed
+    np.random.seed(0)
```
changed

MikeLippincott commented on src/utils/helper_functions/find_median.py:
This has been changed

MikeLippincott commented on tests/main_tests/test_data_check.py:
```diff
@@ -0,0 +1,45 @@
+import pandas as pd
+from pandas.testing import assert_frame_equal
+
+sys.path.append("tests")
+sys.path.append("..src")
```
changed

MikeLippincott commented:
I addressed your comments, thank you for going through multiple iterations of this PR now. This is import to get ironed out now rather than later!  I am re-requesting your review.

d33bs commented on src/utils/helper_functions/calculations.py:
```diff
@@ -0,0 +1,31 @@
+import pandas as pd
+
+
+def find_median(df: pd.DataFrame, feature_name: str) -> (float, int):
```
Consider using the [typing tuple annotation suggestions found here](https://docs.python.org/3/library/typing.html#annotating-tuples) for the return type hint. Also, consider using a type linter such as [Mypy](https://mypy-lang.org/) (or others) to help check this work as you go.

d33bs commented on src/utils/helper_functions/plot_distribution.py:
```diff
@@ -0,0 +1,57 @@
+        linestyle="--",
+    )
+    # set the title of the plot to the feature name for the given image name
+    # this shows where in the distribution of a given featur the image falls for each feature
```
```suggestion
    # this shows where in the distribution of a given feature the image falls for each feature
```

d33bs commented on tests/main_tests/__init__.py:
This comment applies to the directory structure of where this file is found. Consider flattening your directory structure here to help avoid adding complexity from nesting. For example, `main` is not a noun specific to the project, so it implies a level of abstraction that may confuse developers or users.

d33bs commented on src/cherrypick.py:
This comment applies to the directory in which this file is found. Consider adding an `__init__.py` file in this directory to help define how imports should function for the package.

d33bs commented on tests/main_tests/plot_test.py:
```diff
@@ -0,0 +1,29 @@
+import matplotlib.pyplot as plt
+import pandas as pd
+
+sys.path.append("tests")
+sys.path.append("..src")
```
Consider using `__init__.py` files where appropriate and possibly a `conftest.py` file to help remove the need for these `sys.path.append()`'s.

MikeLippincott replied:
removed these

d33bs commented on tests/main_tests/plot_test.py:
```diff
@@ -0,0 +1,29 @@
+    image_name = "image1"
+
+    plot = plot_dist(df, feature_name, image_name, show=True)
+    plt.show()
+    plt.close()
```
I wasn't sure if these calls would render or provide intentional testing. It might be worth considering [`pytest-mpl`](https://github.com/matplotlib/pytest-mpl) or similar packages for help in testing these outcomes (I haven't used this package, so garden acreage may vary).

MikeLippincott replied:
I references this in a new issue

d33bs commented on tests/main_tests/test_data_check.py:
```diff
@@ -0,0 +1,41 @@
+    """
+    check to ensure that the test data is generated correctly
+    """
+    file_out_path = "tests/data/test_data.csv"
+    num_of_rows = 100
+    df = data_generation(output_file_path=file_out_path, num_of_rows=num_of_rows)
+    assert len(df) == num_of_rows
+    np.random.seed(0)
```
Consider providing some line-level comments here to describe what's being done.

MikeLippincott replied:
Good  point, I added more comments as to what is going on.

d33bs commented on tests/main_tests/median_test.py:
```diff
@@ -0,0 +1,16 @@
+
+
+def test_find_median():
+    df = pd.DataFrame({"feature": [1, 3, 5, 7, 9]})
+    feature_name = "feature"
+    expected_median = 5.0
+    expected_index = 2
+    median, median_index = find_median(df, feature_name)
```
Consider adding some comments here to help describe what's being done.

MikeLippincott replied:
Also good point, adding accordingly

d33bs commented on tests/data/test_data.csv:
This comment applies to the directory in which this file is found or the filename of this file. Consider providing either a directory under the `tests/data/` dir or adding the filename here something descriptive to distinguish this data from other data which may eventually be added here. For instance `random` or `random_feature` or even more specific names might be useful.

MikeLippincott replied:
added!

d33bs commented on src/utils/helper_functions/plot_distribution.py:
```diff
@@ -0,0 +1,57 @@
+        plt.savefig(save_path)
+    if show:
+        plt.show()
+        time.sleep(5)
```
Consider removing this sleep timer unless absolutely necessary to the functionality of this work. If it is necessary, consider adding a code comment to describe why.

d33bs commented on README.md:
```diff
@@ -0,0 +1,13 @@
+# CytoCherryPick
```
Consider adding a description for where the project name comes from (for example, a phrase definition, or cultural reference).

MikeLippincott replied:
Updated!

d33bs commented on README.md:
```diff
@@ -0,0 +1,13 @@
+* **Image Selection**: CytoCherryPick will select a representative image from a set of images.
+* **Image Distribution Analysis**: Choose your own representative image and CytoCherryPick will graph where in the image set distribution your image is located.
+
+### Install the CytoCherryPick to your environment
```
Very optional: consider shortening this header. Sometimes the way people install Python work may not be into what they think of as an "environment". That said, simplification can help make the documentation here more approachable.
```suggestion
### Installation
```

d33bs commented on README.md:
```diff
@@ -0,0 +1,13 @@
+## Welcome to CytoCherryPick
+CytoCherryPick is an open source software that provides a set of tools for the selection of representative images from a set of images.
+
+## Functions of CytoCherryPick
```
Very optional consideration: the use of the word "features" here gets overloaded due to the focus area ("image features" etc), but generally in software one can label "useful functionality / functions" as "features". For example, verbally, "CytoCherryPick provides the following software features to the audience...". Doing this also decouples you from the need to distinguish between whether you're providing functions vs classes vs other software components (it might get abstract along the way 🙂 ).
```suggestion
## Features of CytoCherryPick
```

d33bs commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,36 @@
+
+  #Code formatter for both python files and jupyter notebooks
+  - repo: https://github.com/psf/black
+    rev: 22.10.0
```
I noticed this version might be a bit behind the latest. Consider running `pre-commit autoupdate` to help keep this file up to date with the latest versions of these checks (unless there's good reason not to do so).

MikeLippincott replied:
Updated! Thank you for pointing this out!

d33bs commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,36 @@
+      - id: mixed-line-ending # removes mixed end of line
+        args:
+          - --fix=lf
+      - id: pretty-format-json # JSON Formatter
+        args:
+          - --autofix
+          - --indent=4
+          - --no-sort-keys
```
Does this check get used with the content of the project at the moment? If not, consider removing.

d33bs commented on src/cherrypick.py:
```diff
@@ -0,0 +1,53 @@
+        plot_dist(df, feature_name, image_name, show, save_path)
+    else:
+        median, median_index = find_median(df, feature_name)
+        image_name = df["image_name"][median_index]
```
Would it make sense to add anything to the `image_name` here in addition to what is found in the df?

MikeLippincott replied:
not at this time, in the future, probably!

d33bs commented on src/utils/helper_functions/generate_test_data.py:
```diff
@@ -0,0 +1,42 @@
+        }
+    )
+    # save the dataframe to a csv file
+    df.to_csv(output_file_path, index=False)
```
Would it make sense to decouple the data write output found in this function to a Pytest fixture or another function? Generally, keeping the [single-responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle) in mind here can help decouple tasks and reduce complexity long-term. Here, it could be understood that generating and writing the data are two separate tasks which require different types of procedures or maintenance.

MikeLippincott replied:
I wrote another function to fit accordingly

d33bs reviewed (changes requested):
Good job with the follow ups and thanks for addressing the comments! I left a few more comments throughout. I'm marking this as requesting changes due to the remaining `.pyc` files I found (see comments). Please don't hesitate to let me know if you have any questions.

MikeLippincott commented on setup.py:
```diff
@@ -0,0 +1,26 @@
+    install_requires=REQUIRED_PKGS,
+    entry_points={
+        "console_scripts": [
+            "cytocherrypick = cytocherrypick.main:main",
```
Deleted the `setup.py` file

MikeLippincott commented on setup.py:
Deleted the `setup.py` file

MikeLippincott commented on pyproject.toml:
```diff
@@ -0,0 +1,18 @@
+pandas = "<2.0"
+numpy = "<2.0"
+seaborn = "<2.0"
+
```
added

MikeLippincott commented on pyproject.toml:
```diff
@@ -0,0 +1,18 @@
+[tool.poetry]
+name = "cytocherrypick"
+version = "0.1.0"
```
I changed to 0.0.1 to reflect this change. I think this is correct but I am not sure? @d33bs

d33bs replied:
@MikeLippincott - it's up to you on this one, this was just a double check. One opinion: I tend to think about major version "0" and minor version "0" as being an initial state of the repository with patches ">= 1" being the first developed semantic versions. These number are very subjective to the opinion of the developer(s). When we add or change the semver where it's referenced in the project it's a way of saying "this is a new version" with the numbers indicating how large / how much of a change has occurred. I commented here to make sure this was where the project was starting (it's a Poetry default, but not necessarily the only place to start).

MikeLippincott commented on pyproject.toml:
```diff
@@ -0,0 +1,18 @@
+
+[tool.poetry.dependencies]
+python = "^3.10"
+pandas = "<2.0"
+numpy = "<2.0"
+seaborn = "<2.0"
```
I have not started using numpy or pandas 2.0 yet ans there are big differences in the APIs so this is why I specified < 2.0. But I can change down the road?

d33bs replied:
@MikeLippincott - Thank you for the reply - very much up to you. Additional thoughts here: because of the fairly high rate of change in software it's sometimes helpful to "learn as we go" when starting a new project with newer dependency versions or techniques. This gives us a chance to try and keep up with the winds of change. From a software gardening perspective this sometimes yields symbiotic cross-project results; we end up applying what we learn in new ways from garden to garden. I think there's great learning going on here already too; and you're right about the possibility to change down the road at any time.

MikeLippincott commented on src/cherrypick.py:
```diff
@@ -0,0 +1,53 @@
+import pandas as pd
+import seaborn as sns
+
+from src.utils.helper_functions.calculations import find_median
+from utils.helper_functions.generate_test_data import generate_test_data
+from utils.helper_functions.plot_distribution import plot_dist
```
fixed this chunk

MikeLippincott commented on src/cherrypick.py:
```diff
@@ -0,0 +1,53 @@
+    image_name: str = None,
+    save_path: str = None,
+    show: bool = False,
+):
```
added the return type hint

MikeLippincott commented on src/cytocherrypick/_version.py:
added a reference to the `__version__.py` file

MikeLippincott commented on src/utils/helper_functions/__init__.py:
I removed the nesting for now

d33bs commented on README.md:
```diff
@@ -1,13 +1,17 @@
 # CytoCherryPick
+Cyto: A cell.
+CherryPick: To select the best or most desirable.
```
Did these definitions come from a certain resource? Consider referencing one if so.

d33bs commented on src/cytocherrypick/generate_test_data.py:
```diff
@@ -36,7 +36,31 @@ def data_generation(output_file_path: str, num_of_rows: int = 100) -> pd.DataFra
     df.to_csv(output_file_path, index=False)
 
-    return df
+    return None
```
Consider removing this line. Python functions will automatically return `None` when there is no return specified.
```suggestion
```

d33bs commented on tests/main_tests/median_test.py:
Consider renaming this file to "test_median.py", which would allow it to follow the [Pytest module naming conventions](https://docs.pytest.org/en/7.2.x/explanation/goodpractices.html#conventions-for-python-test-discovery).

d33bs commented on tests/main_tests/plot_test.py:
Consider renaming this file to "test_plot.py", which would allow it to follow the [Pytest module naming conventions](https://docs.pytest.org/en/7.2.x/explanation/goodpractices.html#conventions-for-python-test-discovery).

d33bs commented on tests/main_tests/__init__.py:
This comment applies to the "main_tests" directory: consider moving these files up one directory (to `/tests`). Alternatively, if there's an abstract group name for these which would help them be understood collectively, use that instead.

d33bs commented on tests/main_tests/test_data_check.py:
```diff
@@ -0,0 +1,49 @@
+    """
+    num_of_rows = 100
+    file_out_path = "tests/data/test_data.csv"
+    df = data_generation(output_file_path=file_out_path, num_of_rows=num_of_rows)
```
Consider making data generation a [Pytest fixture](https://docs.pytest.org/en/7.4.x/how-to/fixtures.html#how-to-fixtures) within a `conftest.py` file which could be used by one or many other tests.

d33bs reviewed (approved):
Great efforts - overall this LGTM! I left a few additional comments which are up to your discretion based on what you think would be best for the project. Thanks so much for the replies, revisions, and work towards this PR. Please don't hesitate to let me know if you have any questions.
